src/files/files.ts

```diff
@@ -0,0 +1,250 @@
+/**
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+import { convertToExcalidrawElements } from '@excalidraw/excalidraw'
+import type {
+	BinaryFileData,
+	DataURL,
+	ExcalidrawImperativeAPI,
+} from '@excalidraw/excalidraw/types/types'
+import { Collab } from '../collaboration/collab'
+import type { FileId } from '@excalidraw/excalidraw/types/element/types'
+
+type Meta = {
+	name: string,
+	type: string,
+	lastModified: number,
+	dataurl: string,
+	fileId: string,
+}
+
+export class FileHandle {
+
+	private collab: Collab
+	private excalidrawApi: ExcalidrawImperativeAPI
+	private types: string[]
+	constructor(
+		excalidrawApi: ExcalidrawImperativeAPI,
+		collab: Collab,
+		types: string[],
+	) {
+		this.collab = collab
+		this.excalidrawApi = excalidrawApi
+		this.types = types
+		const containerRef = document.getElementsByClassName(
+			'excalidraw-container',
+		)[0]
+		const constructedFile: BinaryFileData = {
+			mimeType: 'image/png',
+			created: 0o0,
+			id: 'placeholder_image' as FileId,
+			dataURL: 'data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAyNCAyNCI+PHBhdGggZD0iTTE0LDJMMjAsOFYyMEEyLDIgMCAwLDEgMTgsMjJINkEyLDIgMCAwLDEgNCwyMFY0QTIsMiAwIDAsMSA2LDJIMTRNMTgsMjBWOUgxM1Y0SDZWMjBIMThNMTIsMTlMOCwxNUgxMC41VjEySDEzLjVWMTVIMTZMMTIsMTlaIiAvPjwvc3ZnPg==' as DataURL,
+		}
+		this.collab.addFile(constructedFile)
+		if (containerRef) {
+			containerRef.addEventListener('drop', (ev) =>
+				this.filesDragEventListener(ev),
+			)
+		}
+		let lastPointerDown = 0
+		// eslint-disable-next-line @typescript-eslint/no-unused-vars
+		this.excalidrawApi.onPointerDown((activeTool, state, event) => {
+			const clickedElement = state.hit.element
+			if (!clickedElement || !clickedElement.customData) {
+				return
+			}
+			event.stopPropagation()
+			if (Date.now() - lastPointerDown > 200) {
+				lastPointerDown = Date.now()
+				return
+			} else {
+				lastPointerDown = Date.now()
+			}
+			this.downloadFile(clickedElement.customData.meta)
```

Comment:
I'm not sure if double click is what we should do here, was rather confusing to me in the first run. We talked about a popup to download the file, maybe that is something to reconsider when we cannot easily have a nice button click behaviour.

Could also be a small bar at the bottom maybe if the "file node" is selected similar to https://m1.material.io/components/snackbars-toasts.html# with a dedicated Download button?

Reply:
I think that the popup is the best option for now. The snack bar is great for mobile devices but I'm worried that it wouldn't work well on larger displays.

Reply:
@marcoambrosini 
I've now implemented it using the toast messages we already have in nextcloud when you click on them the file is downloaded... I don't think it's the best solution, but we can't draw entirely over the canvas with a popup or something because the file-elements are not movable anymore if something is in front of them. Do you think we should design a small new popup or stick with the nextcloud toast message as in the screenshot?

![image](https://github.com/user-attachments/assets/0a3abd18-b8e4-41cf-b687-4a46ac058577)